File: Sources/SnapLingo/ProviderRegistry.swift
import Foundation

enum ProviderKind: String, CaseIterable, Identifiable, Decodable {
    case openAI
    case anthropic
    case gemini
    case zhipuGLM
    case kimi
    case minimax
    case aliyunBailian
    case volcengineArk

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .openAI: return "OpenAI"
        case .anthropic: return "Anthropic"
        case .gemini: return "Google Gemini"
        case .zhipuGLM: return "Zhipu GLM"
        case .kimi: return "Kimi"
        case .minimax: return "MiniMax"
        case .aliyunBailian: return "Alibaba Bailian"
        case .volcengineArk: return "Volcengine Ark"
        }
    }

    var credentialKey: CredentialKey {
        switch self {
        case .openAI: return .openAI
        case .anthropic: return .anthropic
        case .gemini: return .gemini
        case .zhipuGLM: return .zhipuGLM
        case .kimi: return .kimi
        case .minimax: return .minimax
        case .aliyunBailian: return .aliyunBailian
        case .volcengineArk: return .volcengineArk
        }
    }

    var apiKeyPlaceholder: String {
        switch self {
        case .openAI: return "OpenAI API key"
        case .anthropic: return "Anthropic API key"
        case .gemini: return "Gemini API key"
        case .zhipuGLM: return "Zhipu GLM API key"
        case .kimi: return "Kimi API key"
        case .minimax: return "MiniMax API key"
        case .aliyunBailian: return "DashScope API key"
        case .volcengineArk: return "ARK API key"
        }
    }
}

enum ProviderProtocolStyle: String, Decodable {
    case openAIResponses
    case openAIChatCompletions
    case anthropicMessages
    case geminiGenerateContent
}

struct ProviderPreset {
    let kind: ProviderKind
    let style: ProviderProtocolStyle
    let baseURL: String
    let model: String
}

final class ProviderRegistry {
    private let providerCatalog: ProviderCatalog
    private let credentialStore: CredentialStore
    private let session: URLSession
    private let defaults: UserDefaults
    private let providerDefaultsKey = "selected-provider"

    var selectedProvider: ProviderKind {
        get { ProviderKind(rawValue: defaults.string(forKey: providerDefaultsKey) ?? "") ?? .openAI }
        set { defaults.set(newValue.rawValue, forKey: providerDefaultsKey) }
    }

    init(
        providerCatalog: ProviderCatalog = .shared,
        credentialStore: CredentialStore,
        session: URLSession = .shared,
        defaults: UserDefaults = .standard
    ) {
        self.providerCatalog = providerCatalog
        self.credentialStore = credentialStore
        self.session = session
        self.defaults = defaults
    }

    @MainActor
    func currentClient() -> any ProviderClient {
        makeClient(for: selectedProvider)
    }

    @MainActor
    func makeClient(for provider: ProviderKind) -> any ProviderClient {
        let preset = providerPreset(for: provider)

        switch preset.style {
        case .openAIResponses:
            return OpenAICompatibleResponsesProvider(
                preset: preset,
                credentialStore: credentialStore,
                session: session
            )
        case .openAIChatCompletions:
            return OpenAICompatibleChatProvider(
                preset: preset,
                credentialStore: credentialStore,
                session: session
            )
        case .anthropicMessages:
            return AnthropicMessagesProvider(
                preset: preset,
                credentialStore: credentialStore,
                session: session
            )
        case .geminiGenerateContent:
            return GeminiGenerateContentProvider(
                preset: preset,
                credentialStore: credentialStore,
                session: session
            )
        }
    }

    func providerPreset(for provider: ProviderKind) -> ProviderPreset {
        providerCatalog.preset(for: provider)
    }

    func loadKey(for provider: ProviderKind) -> String {
        (try? credentialStore.loadSecret(for: provider.credentialKey)) ?? ""
    }

    func saveKey(_ secret: String, for provider: ProviderKind) throws {
        try credentialStore.saveSecret(secret, for: provider.credentialKey)
    }

    func deleteKey(for provider: ProviderKind) throws {
        try credentialStore.deleteSecret(for: provider.credentialKey)
    }
}

File: SnapLingoWindows/Services/ProviderRegistry.cs
namespace SnapLingoWindows.Services;

public sealed class ProviderRegistry
{
    private readonly SecureSecretStore secretStore;
    private readonly HttpClient httpClient;

    public ProviderRegistry(SecureSecretStore secretStore, HttpClient? httpClient = null)
    {
        this.secretStore = secretStore;
        this.httpClient = httpClient ?? new HttpClient();
    }

    public ProviderKind SelectedProvider { get; set; } = ProviderKind.OpenAI;

    public IProviderClient CurrentClient() => MakeClient(SelectedProvider);

    public IProviderClient MakeClient(ProviderKind provider)
    {
        var preset = GetPreset(provider);

        return preset.Style switch
        {
            ProviderProtocolStyle.OpenAIResponses => new OpenAIResponsesProvider(preset, secretStore, httpClient),
            ProviderProtocolStyle.OpenAIChatCompletions => new OpenAIChatProvider(preset, secretStore, httpClient),
            ProviderProtocolStyle.AnthropicMessages => new AnthropicMessagesProvider(preset, secretStore, httpClient),
            ProviderProtocolStyle.GeminiGenerateContent => new GeminiGenerateContentProvider(preset, secretStore, httpClient),
            _ => throw new InvalidOperationException($"Unsupported provider style: {preset.Style}"),
        };
    }

    public ProviderPreset GetPreset(ProviderKind provider) => provider switch
    {
        ProviderKind.OpenAI => new ProviderPreset(provider, ProviderProtocolStyle.OpenAIResponses, "https://api.openai.com/v1", "gpt-4.1-mini"),
        ProviderKind.Anthropic => new ProviderPreset(provider, ProviderProtocolStyle.AnthropicMessages, "https://api.anthropic.com/v1", "claude-sonnet-4-20250514"),
        ProviderKind.Gemini => new ProviderPreset(provider, ProviderProtocolStyle.GeminiGenerateContent, "https://generativelanguage.googleapis.com/v1beta/models", "gemini-2.5-flash"),
        ProviderKind.ZhipuGLM => new ProviderPreset(provider, ProviderProtocolStyle.OpenAIChatCompletions, "https://api.z.ai/api/paas/v4", "glm-4.5-air"),
        ProviderKind.Kimi => new ProviderPreset(provider, ProviderProtocolStyle.OpenAIChatCompletions, "https://api.moonshot.cn/v1", "kimi-k2-turbo-preview"),
        ProviderKind.MiniMax => new ProviderPreset(provider, ProviderProtocolStyle.OpenAIChatCompletions, "https://api.minimaxi.com/v1", "MiniMax-M2.5-highspeed"),
        ProviderKind.AlibabaBailian => new ProviderPreset(provider, ProviderProtocolStyle.OpenAIChatCompletions, "https://dashscope.aliyuncs.com/compatible-mode/v1", "qwen3.5-flash"),
        ProviderKind.VolcengineArk => new ProviderPreset(provider, ProviderProtocolStyle.OpenAIChatCompletions, "https://ark.cn-beijing.volces.com/api/v3", "doubao-seed-1-6-250615"),
        _ => throw new InvalidOperationException($"Unsupported provider: {provider}"),
    };

    public string LoadKey(ProviderKind provider) => secretStore.LoadSecret(provider);

    public void SaveKey(string secret, ProviderKind provider) => secretStore.SaveSecret(secret, provider);

    public void DeleteKey(ProviderKind provider) => secretStore.DeleteSecret(provider);
}
